### PR TITLE
[fetch_binary_drivers] new [fetch_drivers] package

### PR DESCRIPTION
**Pull request by moriarty**
This is a new package providing the "fetch_drivers" which was previously
built internally and distributed as a binary only debian package.

We will officially announce when the upgrade from Indigo->Melodic is
ready. _**This is not yet the announcement**_.

18.04 and ROS Melodic on the Fetch Research Platforms is still being
tested.

The process isn't straight foward as there are calibration files
which you will want to back-up before the upgrade.

This `fetch_drivers` package also won't start like it use to because of
the change from `upstart` to `systemd`.
That requires `fetch_system_config` which will be open sourced soon.

This `fetch_binary_drivers` package is part of a plan to distribute
updates and upgrades to Fetch Research Platform customers faster in the future.

This relates to fetchrobotics/fetch_ros#63

1. An officiall announcement will be made when everything is ready.
2. We will update and document the process on:
    https://docs.fetchrobotics.com/
3. We will also announce via our mailing list, ros discourse and post on:
    https://opensource.fetchrobotics.com/

### CODE REVIEW

File: fetch_binary_drivers/CMakeLists.txt
Comment (moriarty) [inline fetch_binary_drivers/CMakeLists.txt]: TODO: the internal driver CATKIN_DEPENDS are slightly different between this version and the internal version (for example: no sick microscan3 safety scanners) so I need to sort through the list and include only what we need.

File: fetch_binary_drivers/CMakeLists.txt
Comment (moriarty) [inline fetch_binary_drivers/CMakeLists.txt]: I'm sure there are system dependencies. the binary driver is a shared object not a static one.

**Review by moriarty — COMMENTED** [review]
I will add additional dependencies just to be safe

File: fetch_binary_drivers/README.md
Comment (cjds) [inline fetch_binary_drivers/README.md:43]: Link here

File: fetch_binary_drivers/package.xml
Comment (cjds) [inline fetch_binary_drivers/package.xml:13]: Are these comments intentionally left in

File: fetch_system_config/debian/compat
Comment (cjds) [inline fetch_system_config/debian/compat:1]: You probably want 11?
```
Package: debhelper (9.20131227ubuntu1)
```
even for 14.04

File: fetch_system_config/debian/control
Comment (cjds) [inline fetch_system_config/debian/control:18]: What does this mean?

File: fetch_system_config/root/lib/udev/rules.d/99-logitechF710.rules
Comment (cjds) [inline fetch_system_config/root/lib/udev/rules.d/99-logitechF710.rules:1]: Why are we adding these rules unless we are giving them a logitech?

File: fetch_system_config/root/etc/pm/power.d/wireless
Comment (cjds) [inline fetch_system_config/root/etc/pm/power.d/wireless]: Oh boy!!!

File: fetch_binary_drivers/CMakeLists.txt
Comment (cjds) [inline fetch_binary_drivers/CMakeLists.txt:60]: I can't find a reference in the drivers to auto_dock

File: fetch_binary_drivers/Makefile.tarball
Comment (cjds) [inline fetch_binary_drivers/Makefile.tarball:3]: More specific?

File: fetch_binary_drivers/Makefile.tarball
Comment (cjds) [inline fetch_binary_drivers/Makefile.tarball:5]: Why not use DRIVER_VERSION

File: fetch_binary_drivers/Makefile.tarball
Comment (cjds) [inline fetch_binary_drivers/Makefile.tarball:6]: Ditto as above

File: fetch_binary_drivers/Makefile.tarball
Comment (cjds) [inline fetch_binary_drivers/Makefile.tarball:1]: Why non-standard name?

File: fetch_binary_drivers/README.md
Comment (cjds) [inline fetch_binary_drivers/README.md:10]: Link

File: fetch_binary_drivers/README.md
Comment (cjds) [inline fetch_binary_drivers/README.md:5]: Link

File: fetch_binary_drivers/README.md
Comment (cjds) [inline fetch_binary_drivers/README.md]: ROS Melodic

File: fetch_binary_drivers/README.md
Comment (cjds) [inline fetch_binary_drivers/README.md:33]: Do we need to tell them this?

File: fetch_system_config/debian/fetch-melodic-config.ros.logrotate
Comment (cjds) [inline fetch_system_config/debian/fetch-melodic-config.ros.logrotate]: indigo?

File: fetch_system_config/debian/fetch-melodic-config.ros.logrotate
Comment (cjds) [inline fetch_system_config/debian/fetch-melodic-config.ros.logrotate]: indigo?

File: fetch_binary_drivers/package.xml
Comment (cjds) [inline fetch_binary_drivers/package.xml:4]: Are we not syncing version numbers with the drivers themselves?

File: fetch_binary_drivers/package.xml
Comment (cjds) [inline fetch_binary_drivers/package.xml]: `https://wiki.ros.org/fetch_drivers` for extra security

File: fetch_binary_drivers/package.xml
Comment (cjds) [inline fetch_binary_drivers/package.xml]: Fetch Robotics

File: .gitignore
Comment (cjds) [inline .gitignore:14]: files? what does this do? who uses this folder?

File: fetch_system_config/debian/fetch-melodic-config.postinst
Comment (cjds) [inline fetch_system_config/debian/fetch-melodic-config.postinst]: What's the story with this TODO. Is it done? https://wiki.debian.org/ConfigurePowerButton

File: fetch_system_config/debian/fetch-melodic-config.postinst
Comment (cjds) [inline fetch_system_config/debian/fetch-melodic-config.postinst:69]: Do eth0 and 1 apply on ubuntu 18.04

File: fetch_system_config/debian/fetch-melodic-config.postinst
Comment (cjds) [inline fetch_system_config/debian/fetch-melodic-config.postinst:86]: eth0 1?

File: fetch_system_config/root/opt/ros/roscore_poststart.bash
Comment (cjds) [inline fetch_system_config/root/opt/ros/roscore_poststart.bash:2]: Put the echo in the while so we know it hasn't locked up maybe?

File: fetch_binary_drivers/package.xml
Comment (cjds) [inline fetch_binary_drivers/package.xml:49]: todo: ensure that these are the same as the ones we used when building the binary drivers.

File: fetch_binary_drivers/package.xml
Comment (erelson) [inline fetch_binary_drivers/package.xml]: typo 'wich'

File: fetch_binary_drivers/package.xml
Comment (erelson) [inline fetch_binary_drivers/package.xml]: typo frieght

**Review by erelson — COMMENTED** [review]
Couple typos I found

**Review by cjds — CHANGES_REQUESTED** [review]
First pass. Will do a more thorough review soon

File: fetch_binary_drivers/package.xml
Comment (mikeferguson) [reply fetch_binary_drivers/package.xml:49]: you won't be able to do that -- the ROS buildfarm / repos only contain the latest debs - once a new version is synced, the old ones go away. Hardcoding the value would make your package uninstallable at best.

File: fetch_system_config/debian/fetch-melodic-config.postinst
Comment (erelson) [reply fetch_system_config/debian/fetch-melodic-config.postinst:69]: It is for our robots because I found the computer hardware we use is not 100% guaranteed to have the same e.g. ensp1s0 style names every time it boots. Therefore I implement a "standard" way to disable this newer naming behavior.  See the commit message as well.  I'll also being including notes about this networking decision in the docs for 14.04->18.04.

File: fetch_system_config/debian/fetch-melodic-config.postinst
Comment (erelson) [reply fetch_system_config/debian/fetch-melodic-config.postinst]: It's a non-impactful TODO that I'm hoping to get to this week.

File: fetch_system_config/root/lib/udev/rules.d/99-logitechF710.rules
Comment (erelson) [reply fetch_system_config/root/lib/udev/rules.d/99-logitechF710.rules:1]: This rule was already in there. We've enabled logitech controller usage historically and so it seems reasonable to continue doing so. (I would need to check if research stack has the teleop changes needed for this to work though)

File: fetch_system_config/debian/control
Comment (erelson) [reply fetch_system_config/debian/control:18]: Very brief vague statement, but nominally accurate. It installs roscore + robot systemd units which run at computer power-on. I'll reword this probably.

File: fetch_system_config/debian/fetch-melodic-config.ros.logrotate
Comment (erelson) [reply fetch_system_config/debian/fetch-melodic-config.ros.logrotate]: Yup, gotta fix this! I left these logrotate rules out initially, then re-added them long after my earlier s/indigo/melodic pass.

File: .gitignore
Comment (erelson) [reply .gitignore:14]: it's a file created by dpkg-debbuilder or whatever. I'll add a better comment to the .gitignore file

File: fetch_system_config/debian/compat
Comment (erelson) [reply fetch_system_config/debian/compat:1]: My 14.04 desktop (still on 3.x kernel... nice and safe...) has 9.x debhelper installed. I don't think there's a reason to bump this requirement presently. And it sounds like the plan is we'll move to using bloom for this later anyways.

File: fetch_system_config/root/etc/pm/power.d/wireless
Comment (erelson) [reply fetch_system_config/root/etc/pm/power.d/wireless]: I need to reinvestigate this.

File: fetch_system_config/root/opt/ros/roscore_poststart.bash
Comment (erelson) [reply fetch_system_config/root/opt/ros/roscore_poststart.bash:2]: I'm going to leave this the way it was for simplicity right now.

File: fetch_binary_drivers/README.md
Comment (erelson) [reply fetch_binary_drivers/README.md:5]: Markdown will render these fine. Is this a style preference?

File: fetch_binary_drivers/package.xml
Comment (moriarty) [reply fetch_binary_drivers/package.xml:49]: @mikeferguson: would a nightly build job of our binaries, ensuring we’ve always got the latest dependencies work here?

@cjds and I were hoping to open source the Dockerfile & Python script which builds the `output.tar.gz`- it won’t be helpful without pull access on our drivers but it would at least make the process transparent

Comment (mikeferguson) [reply fetch_binary_drivers/package.xml:49]: @moriarty even having the released deb download a nightly binary build is a challenge -- when a sync to public happens, you'd have a ~24 hour window where the version built against public is no longer right. (and that sync happens every ~ 2 weeks, so that's about 7% downtime). Also how do you make it work on both shadow-fixed (where developers tend to test) and the main ros repo?

Comment (moriarty) [reply fetch_binary_drivers/package.xml:49]: :disappointed:
@dbking77 & @chadrockey maybe long term we reopen the discussion about refactoring some things to better support public/private drivers to make releasing to research customers easier :question: 

We might need a second version setup up and built against shadow-fixed, this still doesn't fix the original problem, and it also wouldn't be clear how to set that up so that the shadow-fixed pulled a `fetch-drivers-shadow.tar.gz` and the release pulled `fetch-drivers.tar.gz` especially since currently they're pulling: `http://packages.fetchrobotics.com/binaries/fetch-drivers-0.8.0.tar.gz`
... Fetch3 & my machine are currently setup against shadow-fixed, but these drivers were built against whatever is inside of

```dockerfile
FROM ros:melodic-robot-bionic
```
(As mentioned @cjds will cleanup the creation process and we plan to put them:
https://github.com/fetchrobotics/fetch_firmware_repackager or rename that repository `fetch_drivers_repackager`)

And it works now, but that's likely because there hasn't been any API/ABI breaking changes, which is not always guaranteed.
We'll need to check which of the dependencies we have are core dependencies, and which are non-core dependencies. 

For the most part, the 7% downtime is a worst case calculation coming from 24hr every 2 weeks. If the dependencies are part of the core ros repositories, I think the API/ABI is stable enough or at least suppose to be... For non-core, and not released by us then it's just a risk that we'll probably need to take.
I don't even know if the melodic sync is triggered every 2 weeks because most of the folks at open robots are developing towards ros2. @jacobperron & @wjwwood?

Our plan, was to setup a mirror of ROS, which we manually trigger a sync on.
So customers with Fetch Research Robots could point to our http://packages.fetchrobotics.com/ and get guaranteed stability. But have the option to get the packages from packages.ros.org either ros or ros-shadow-fixed if they really want to live on the edge and accept some instability.

By pushing this into the public build farm, our internal `packages.fetchrobotics.com` will become almost a pure mirror, which should have much maintainable for our DevOps team and @bluryi.

If we sync our mirror manually, after testing to know we're not in the 7% downtime, then we'll get stability and more frequent updates.

For the 7% of possible down time for the `packages.ros.org` I think @erelson will need to update the (not yet existing) documentation to include strong warnings about pointing to `packages.fetchrobotics.com` or `packages.ros.org`
The packages.fetchrobotics.com mirror for *melodic* does not yet exist, we were going to quietly get our packages into melodic, and then setup a mirror once they were in and synced.

Comment (wjwwood) [reply fetch_binary_drivers/package.xml:49]: @clalancette is the ROS Boss for Melodic, so it's up to him to decide when the syncs happen, though I think they've been trying to keep a regular cadence. Looking at the discourse posts, it's more like once a month, though maybe that will change.

Perhaps you can anticipate the syncs and be prepared with updated packages of your own to sync in order to minimize the period where your debs are out of date. We do usually announce an intention to sync a week or two before doing it, with the intention that people could test it for unintended regressions.

----

I don't know if this would be relevant as well, but there's this recent announcement by @nuclearsandwich:

https://discourse.ros.org/t/announcing-ros-snapshot-repositories/7705

Comment (clalancette) [reply fetch_binary_drivers/package.xml:49]: My intention is to sync Melodic every two weeks, though I've basically never been able to achieve that due to regressions that I have to fix for every sync.  So the cadence has indeed been more like once a month.

I intend to continue to attempt to sync every two weeks unless there is a compelling reason to change that.  So if you have reasons you'd like it to change, please let me know and we can consider it.

File: fetch_system_config/root/etc/pm/power.d/wireless
Comment (erelson) [reply fetch_system_config/root/etc/pm/power.d/wireless]: This is supposed to turn of power managment of the wifi card, and works in 14.04 (checked via output of `iwconfig` but isn't working on Fetch1 with 18.04.

Looks like the new way to make this work, (and verified) is to change wifi.powersave = 3 to = 2 in /etc/NetworkManager/conf.d/default-wifi-powersave-on.conf :+1:

File: fetch_system_config/debian/fetch-melodic-config.postinst
Comment (erelson) [inline fetch_system_config/debian/fetch-melodic-config.postinst:126]: fyi @cjds here's the updated approach :+1:

File: fetch_binary_drivers/package.xml
Comment (moriarty) [reply fetch_binary_drivers/package.xml:13]: They're just left over from the catkin generate package command. I think it was just reminding me to add the maintainers.

File: fetch_binary_drivers/CMakeLists.txt
Comment (moriarty) [reply fetch_binary_drivers/CMakeLists.txt:60]: @cjds it might be stripped out of the research platform drivers.

```
amoriarty at SW44 in ~/ros/melodic/fetch-private/src/fetch_drivers on master
$ grep -rl auto_dock_msgs .
./CMakeLists.txt
./include/fetch_drivers/logpro/client.h
./package.xml
./src/logpro_client.cpp
```

File: fetch_binary_drivers/Makefile.tarball
Comment (moriarty) [reply fetch_binary_drivers/Makefile.tarball:3]: There are too many to list. I had several TODOs and I just combined them into one generic one.
I see your comments also say `Why not use DRIVER_VERSION` and that's basically when I rolled back and just used the hardcoded path for now.
But it goes into the comments on another comment were we'll need to setup more build/test automation.

For now we're just pulling `DRIVER_VERSION = 0.8.0` and until we have a more automated pipeline we'll leave it at this version and just be sure to rebuild that version of our drivers against the latest melodic especially at the time of a sync.

**Review by cjds — APPROVED** [review]
LGTM. This works for a 0.1 release. We'll be improving it over time.

**Comment by moriarty** [comment]
Once https://github.com/ros/rosdistro/pull/20301 is in the rosdep keys for `fetch_description` and `fetch_teleop` should be available and then I can bloom this.

**Comment by moriarty** [comment]
https://github.com/ros/rosdistro/pull/20303

@cjds and @erelson we'll test this once on hardware before we merge the PR to the documentation website which officially announces melodic upgrade process.

File: fetch_binary_drivers/package.xml
Comment (moriarty) [reply fetch_binary_drivers/package.xml:49]: @nickswalker and @mikeferguson I wanted to give you a warning: just in case you've been running melodic and testing with Fetch's using **ros-shadow-fixed** this is now there:

http://repositories.ros.org/status_page/ros_melodic_default.html?q=fetch_

@erelson will do a test on hardware using the **ros-shadow-fixed** before we make an official announcement with links to upgrade documentation. So: best case, everything works, but it's more likely that the official announcement to the fetch user mailing lists is going to be not the upcoming ros melodic sync, but the one after. If we're really unlucky we could be in one of those 7% downtime scenarios which was mentioned above.

**Pull request merged**
